File: decine_cpp2/src/DeCine.cpp
#include "DeCine.h"

namespace decine {

DeCine *DeCine::instancia = new DeCine();

	DeCine *DeCine::getDeCine() {
		return instancia;
	}

	DeCine::DeCine() {
		InitializeInstanceFields();
	}

	UsuarioVIP DeCine::buscarUsuario(const std::string &correo) throw(std::exception) {

		for (auto usuario: misUsuarios) {
			if (usuario.getCorreo() == correo) {
				return usuario;
			}
		}

		throw std::runtime_error("no existe el usuario con ese correo");

	}

	vector<string> DeCine::obtenerUsuariosVIP() {

		vector<string> resultado;
		for(auto value: misUsuarios) {
			resultado.push_back(value.getCorreo());
		}
		return resultado;
	}

	vector<string> DeCine::obtenerPeliculas() {
		vector<string> resultado;
		for(auto value: misPeliculas) {
			resultado.push_back(value.getTitulo());
		}
		return resultado;
	}

	vector<string> DeCine::obtenerActores() {
		vector<string> resultado;
		for(auto value: misActores) {
			resultado.push_back(value.getNombre());
		}
		return resultado;	
	}

	vector<string> DeCine::obtenerDirectores() {
		vector<string> resultado;
		for(auto value: misDirectores) {
			resultado.push_back(value.getNombre());
		}
		return resultado;	
	}

	vector<string> DeCine::obtenerPremios() {
		vector<string> resultado;
		for(auto value: misFestivales) {
			resultado.push_back(value.getNombre());
		}
		return resultado;	
	}

	void DeCine::incluirActor(const std::string &nombre, const std::string &nacionalidad, const std::string &web) {
		misActores.push_back(Actor(nombre,nacionalidad, web));
	}

	void DeCine::incluirDirector(const std::string &nombre, const std::string &nacionalidad, const std::string &web) {

		Director director = Director(nombre,nacionalidad, web);

		misDirectores.push_back(director);
	}

	void DeCine::incluirFestivalPremio(const std::string &nombre) {
		misFestivales.push_back(FestivalPremio(nombre));
	}

	void DeCine::incluirUsuario(const std::string &nombre, const std::string &correo) {
	   misUsuarios.push_back(UsuarioVIP(nombre,correo));
	}

	void DeCine::incluirPelicula(const std::string &titulo, Pelicula &pelicula) {
		misPeliculas.push_back(pelicula);
	}

	void DeCine::nuevaPelicula(const std::string &titulo, const std::string &tituloOriginal, int añoEstreno, int duracion, const std::string &guion, const std::string &musica, const std::string &pais, const std::string &webOficial, const std::string &productora, const std::string &sinopsis, float recaudacion) throw(std::exception) {

		bool existe = existePelicula(titulo);

		if (existe) {
			throw std::runtime_error("Esa película ya está incluida en el sistema");
		}

		Pelicula miPelicula = Pelicula(titulo, tituloOriginal, añoEstreno, duracion, guion, musica, pais, webOficial, productora, sinopsis, recaudacion);

		incluirPelicula(titulo, miPelicula);



	}

	std::string DeCine::busquedaPelicula(std::vector<std::string> &criterioBusqueda) {


		std::string resultadoBusqueda = "";


		std::vector<Pelicula> peliculas = seleccionarPeliculas(criterioBusqueda);

		for (auto pelicula : peliculas) {

			std::string datosBasicos = pelicula.obtenerDatosBasicos();


			resultadoBusqueda += datosBasicos;

			resultadoBusqueda += std::string("Directores:\n");
			std::vector<Director> directores = pelicula.getDirectores();
			for (auto director : directores) {
				resultadoBusqueda += director.getNombre() + std::string("\n");
			}

			resultadoBusqueda += std::string("Actores:\n");
			std::vector<Actor> actores = pelicula.getActores();
			for (auto actor : actores) {
				resultadoBusqueda += actor.getNombre() + std::string("\n");
			}



			if (StringConverterHelper::fromString<int>(criterioBusqueda[0]) == 1) {


				std::string datosComplementarios = pelicula.obtenerDatosComplementarios();


				// Esta mal en el diagrama
				resultadoBusqueda += datosComplementarios;


			}


		}


		return resultadoBusqueda;

	}

	void DeCine::nuevoGeneroPelicula(const std::string &titulo, Genero miGenero) throw(std::exception) {

		Pelicula miPelicula = buscarPelicula(titulo);

		miPelicula.vincularGenero(miGenero);



	}

	void DeCine::nuevoDirectorPelicula(const std::string &titulo, const std::string &nombreDirector) throw(std::exception) {

		Pelicula miPelicula = buscarPelicula(titulo);

		Director miDirector = buscarDirector(nombreDirector);

		miPelicula.vincularDirector(miDirector);


	}

	void DeCine::nuevoActorPelicula(const std::string &titulo, const std::string &nombreActor) throw(std::exception) {

		Pelicula miPelicula = buscarPelicula(titulo);

		Actor miActor = buscarActor(nombreActor);

		miPelicula.vincularActor(miActor);

	}

	void DeCine::criticaPelicula(const std::string &correo, const std::string &titulo, const std::string &tituloCritica, const std::string &comentario) throw(std::exception) {

		UsuarioVIP uVIP = buscarUsuarioVIP(correo);

		Pelicula peli = buscarPelicula(titulo);

		peli.incluirCritica(uVIP, tituloCritica, comentario);


	}

	void DeCine::nuevoPremio(const std::string &titulo, const std::string &nombreFP, const std::string &categoria, int año) throw(std::exception) {

		Pelicula pelicula = buscarPelicula(titulo);

		FestivalPremio premio = buscarPremio(nombreFP);

		pelicula.incluirPremio(premio, categoria, año);


	}

	std::vector<Pelicula> DeCine::seleccionarPeliculas(std::vector<std::string> &criterioBusqueda) {

		std::vector<Pelicula> listaPeliculas;


		for (auto pelicula : misPeliculas) {

			int criterio = StringConverterHelper::fromString<int>(criterioBusqueda[0]);
			std::string valor = criterioBusqueda[1];

			switch (criterio) {

				case 1: {// Por titulo
					if (valor == pelicula.getTitulo()) {
						listaPeliculas.push_back(pelicula);
					}

					break;}
				case 2: {// Por director

					std::vector<Director> directores = pelicula.getDirectores();
					for (auto director : directores) {

						if (valor == director.getNombre()) {
							 listaPeliculas.push_back(pelicula);
							 break; // Paramos el bucle (ya ha encontrado el elemento
						}
					}


					break;}
				case 3: {// Por actor

					std::vector<Actor> actores = pelicula.getActores();
					for (auto actor : actores) {

						if (valor == actor.getNombre()) {
							 listaPeliculas.push_back(pelicula);
							 break; // Paramos el bucle (ya ha encontrado el elemento
						}
					}


					break;}
				case 4: {// Por genero

					std::vector<Genero> generos = pelicula.getGeneros();
					for (auto genero : generos) {

						//POR HACER!!!
						// if (genero == valor) {
						// 	listaPeliculas.push_back(pelicula);
						// 	 break; // Paramos el bucle (ya ha encontrado el elemento
					 

					}

					break;}
				case 5: {// Por año estreno

					if (StringConverterHelper::fromString<int>(valor) == pelicula.getAñoEstreno()) {
							 listaPeliculas.push_back(pelicula);


					break;}



					}




			}

		}

		return listaPeliculas;

	}

	bool DeCine::existePelicula(const std::string &titulo) {

		for (auto pelicula: misPeliculas) {
			if (pelicula.getTitulo() == titulo) {
				return true;
			}
		}

		return false;
	}

	Pelicula DeCine::buscarPelicula(const std::string &titulo) throw(std::exception) {

		for (auto pelicula: misPeliculas) {
			if (pelicula.getTitulo() == titulo) {
				return pelicula;
			}
		}

		throw std::runtime_error("no existe la pelicula");

	}

	UsuarioVIP DeCine::buscarUsuarioVIP(const std::string &correo) throw(std::exception) {
		for (auto usuario: misUsuarios) {
			if (usuario.getCorreo() == correo) {
				return usuario;
			}
		}

		throw std::runtime_error("no existe el usuario con ese correo");

	}

	Director DeCine::buscarDirector(const std::string &idDirector) throw(std::exception) {
		for (auto director: misDirectores) {
			if (director.getNombre() == idDirector) {
				return director;
			}
		}

		throw std::runtime_error(std::string("no existe el director con ese id: "));


	}

	Actor DeCine::buscarActor(const std::string &idActor) throw(std::exception) {
		for (auto actor: misActores) {
			if (actor.getNombre() == idActor) {
				return actor;
			}
		}

		throw std::runtime_error(std::string("no existe el actor con ese id: "));
	}

	FestivalPremio DeCine::buscarPremio(const std::string &nombre) throw(std::exception) {

		for (auto festival: misFestivales) {
			if (festival.getNombre() == nombre) {
				return festival;
			}
		}

		throw std::runtime_error("no existe el festival con ese nombre");

	}

	void DeCine::InitializeInstanceFields() {
		// misUsuarios = std::vector();
		// misActores = std::vector();
		// misDirectores = std::vector();
		// misFestivales = std::vector();
		// misPeliculas = std::vector();
	}
}

File: decine_java/src/decine/DeCine.java
/*  
 *  Fundamentos de Ingenieria del Software
 *  Grado en Ingeniería Informática
 * 
 *  2014 © Copyleft - All Wrongs Reserved
 *
 *  Ernesto Serrano <info@ernesto.es>
 *  
 *  
 * 
 */
package decine;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;

public class DeCine { 
    
    // Definición de atributos 
    private Map<String,UsuarioVIP> misUsuarios = new HashMap();
    private Map<String, Actor> misActores = new HashMap();
    private Map <String, Director> misDirectores = new HashMap();
    private Map <String, FestivalPremio> misFestivales = new HashMap();
    private Map <String, Pelicula> misPeliculas = new HashMap();

    // incluir los que faltan
    
    // implementación del Singleton
    private static DeCine instancia = new DeCine();
    public static DeCine getDeCine(){
        return instancia;
    }
    private DeCine(){}
    
    // Búsqueda de un determinado elemento en un Map
    private UsuarioVIP buscarUsuario(String correo) throws Exception{    
        if (!misUsuarios.containsKey(correo)) throw new Exception ("no existe el usuario con ese correo");
        return misUsuarios.get(correo);
    }
    
    // Obtener todos los id (key) de una lista de objetos.
    public Set<String> obtenerUsuariosVIP(){
        return  misUsuarios.keySet();    
    }
    public Set<String> obtenerPeliculas(){
        return  misPeliculas.keySet();    
    }
    public Set<String> obtenerActores(){
        return  misActores.keySet();    
    }
    public Set<String> obtenerDirectores(){
        return  misDirectores.keySet();    
    }
    public Set<String> obtenerPremios(){
        return  misFestivales.keySet();    
    }    
 
     // .... Para todos los demas igual la busqueda de uno concreto o de todas sus claves
   
    // Inicialización del sistema 
    public void incluirActor(String nombre, String nacionalidad, String web){
        misActores.put(nombre, new Actor(nombre,nacionalidad, web));
    }
    public void incluirDirector(String nombre, String nacionalidad, String web){
        misDirectores.put(nombre, new Director(nombre,nacionalidad, web));
    }
    public void incluirFestivalPremio(String nombre){
        misFestivales.put(nombre, new FestivalPremio(nombre));
    }
     public void incluirUsuario(String nombre, String correo){
        misUsuarios.put(correo, new UsuarioVIP(nombre,correo));
    }
     
     void incluirPelicula(String titulo, Pelicula pelicula) {
         misPeliculas.put(titulo, pelicula);
     }
     
     
     // NOTA DURACION ESTA COMO INT Y FLOAT EN LOS DIAGRAMAS
     
    public void nuevaPelicula(String titulo, String tituloOriginal, int añoEstreno, int duracion, String guion, String musica, String pais, String webOficial, String productora, String sinopsis, float recaudacion) throws Exception {

        boolean existe = existePelicula(titulo);
        
        if (existe) throw new Exception ("Esa película ya está incluida en el sistema");
        
        Pelicula miPelicula = new Pelicula(titulo, tituloOriginal, añoEstreno, duracion, guion, musica, pais, webOficial, productora, sinopsis, recaudacion);
        
        incluirPelicula(titulo, miPelicula);
        
    
    
    }
    
    public String busquedaPelicula(ArrayList<String> criterioBusqueda) {
        
        
        String resultadoBusqueda = "";
        
        
        ArrayList<Pelicula> peliculas = seleccionarPeliculas(criterioBusqueda);
        
        for (Pelicula pelicula: peliculas) {
        
            String datosBasicos = pelicula.obtenerDatosBasicos();
            
            
            resultadoBusqueda += datosBasicos;
            
            resultadoBusqueda += "Directores:\n";
            ArrayList<Director> directores = pelicula.getDirectores();            
            for (Director director: directores) {
                resultadoBusqueda += director.getNombre() + "\n";
            }
            
            resultadoBusqueda += "Actores:\n";
            ArrayList<Actor> actores = pelicula.getActores();
            for (Actor actor: actores) {
                resultadoBusqueda +=  actor.getNombre() + "\n";
            }
            
            
            
            if (Integer.parseInt(criterioBusqueda.get(0)) == 1) {
            
            
                String datosComplementarios = pelicula.obtenerDatosComplementarios();
                
                
                // Esta mal en el diagrama
                resultadoBusqueda += datosComplementarios;
                
            
            }

            
        }
        
        
        return resultadoBusqueda;
        
    }
    
    public void nuevoGeneroPelicula(String titulo, Genero miGenero) throws Exception {
    
        Pelicula miPelicula = buscarPelicula(titulo);
     
        miPelicula.vincularGenero(miGenero);

        
    
    }
    
    public void nuevoDirectorPelicula(String titulo, String nombreDirector) throws Exception {
        
        Pelicula miPelicula = buscarPelicula(titulo);

        Director miDirector = buscarDirector(nombreDirector);
        
        miPelicula.vincularDirector(miDirector);

        
    }

    public void nuevoActorPelicula(String titulo, String nombreActor) throws Exception {
        
        Pelicula miPelicula = buscarPelicula(titulo);

        Actor miActor = buscarActor(nombreActor);
        
        miPelicula.vincularActor(miActor);  
        
    }
    
    public void criticaPelicula(String correo, String titulo, String tituloCritica, String comentario) throws Exception {
        
        UsuarioVIP uVIP = buscarUsuarioVIP(correo);
        
        Pelicula peli = buscarPelicula(titulo);
        
        peli.incluirCritica(uVIP, tituloCritica, comentario);

        
    }
    
    public void nuevoPremio(String titulo, String nombreFP, String categoria, int año) throws Exception {
        
        Pelicula pelicula = buscarPelicula(titulo);
        
        FestivalPremio premio = buscarPremio(nombreFP);
        
        pelicula.incluirPremio(premio, categoria, año);
        
        
    }
    
    private ArrayList<Pelicula> seleccionarPeliculas(ArrayList<String> criterioBusqueda) {
        
        ArrayList<Pelicula> listaPeliculas = new ArrayList();
        

        
        for(Entry<String, Pelicula> entry : misPeliculas.entrySet()) {
            
            int criterio = Integer.parseInt(criterioBusqueda.get(0));
            String valor = criterioBusqueda.get(1);
            Pelicula pelicula = entry.getValue();
            
            switch (criterio) {
            
                case 1: // Por titulo
                    if (valor.equals(pelicula.getTitulo()))
                        listaPeliculas.add(pelicula);

                    break;
                case 2: // Por director
                    
                    ArrayList<Director> directores = pelicula.getDirectores();
                    for (Director director: directores) {
                    
                        if (valor.equals(director.getNombre())) {
                             listaPeliculas.add(pelicula);
                             break; // Paramos el bucle (ya ha encontrado el elemento
                        }
                    }
                    
                    
                    break;
                case 3: // Por actor
                    
                    ArrayList<Actor> actores = pelicula.getActores();
                    for (Actor actor: actores) {
                    
                        if (valor.equals(actor.getNombre())) {
                             listaPeliculas.add(pelicula);
                             break; // Paramos el bucle (ya ha encontrado el elemento
                        }
                    }
                    
                    
                    break;
                case 4: // Por genero
                    
                    ArrayList<Genero> generos = pelicula.getGeneros();
                    for (Genero genero: generos) {
                    
                        if (genero.equals(Genero.valueOf(valor))) {
                            listaPeliculas.add(pelicula);
                             break; // Paramos el bucle (ya ha encontrado el elemento
                        }
                        
                    }
                    
                    break;
                case 5: // Por año estreno
                    
                    if (Integer.parseInt(valor) == pelicula.getAñoEstreno()) {
                             listaPeliculas.add(pelicula);
                    
                    
                    break;
                    
            
            
            }
                 
           
                
                
            }
            
        }
       
        return listaPeliculas;
        
    }
    
    private boolean existePelicula(String titulo) throws Exception {
         
        return misPeliculas.containsKey(titulo);
    }
    
    private Pelicula buscarPelicula(String titulo) throws Exception {

        if (!misPeliculas.containsKey(titulo)) throw new Exception ("no existe la pelicula id");
        return misPeliculas.get(titulo);
    }
    
    private UsuarioVIP buscarUsuarioVIP(String correo) throws Exception {
        if (!misUsuarios.containsKey(correo)) throw new Exception ("no existe el usuario con ese correo");
        return misUsuarios.get(correo);
    }
    
    private Director buscarDirector(String idDirector) throws Exception {
        
        if (!misDirectores.containsKey(idDirector)) throw new Exception ("no existe el director con ese id: " + idDirector);
        return misDirectores.get(idDirector);
        
    }
    
    private Actor buscarActor(String idActor) throws Exception {
        if (!misActores.containsKey(idActor)) throw new Exception ("no existe el actor con ese id");
        return misActores.get(idActor);
    }
    
    private FestivalPremio buscarPremio(String nombre) throws Exception {
        if (!misFestivales.containsKey(nombre)) throw new Exception ("no existe el festival con ese nombre");
        return misFestivales.get(nombre);
    }
     
     
}
